tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java

```diff
@@ -1023,4 +1032,60 @@ public void testIllegalManagedLedger() throws Exception {
         verifySendMessageToPartition(topicPartition, newNormalRecords(), Errors.NOT_LEADER_FOR_PARTITION, -1L);
         verifySendMessageToPartition(topicPartition, newAbortTxnMarker(), Errors.NOT_LEADER_FOR_PARTITION, -1L);
     }
+
+
+    /**
+     * Test the sending speed of fetch request when the readable data is less than fetch.minBytes.
+     */
+    @Test(timeOut = 60000)
+    public void testFetchMinBytesSingleConsumer() throws Exception {
+        String topicName = "testMinBytesTopic";
+        TopicPartition tp = new TopicPartition(topicName, 0);
+
+        // create partitioned topic.
+        admin.topics().createPartitionedTopic(topicName, 1);
+        List<TopicPartition> topicPartitions = new ArrayList<>();
+        topicPartitions.add(tp);
+
+        int maxWaitMs = 3000; // very long time
+        int minBytes = 1;
+        // case1: consuming an empty topic.
+        @Cleanup
+        KafkaConsumer<String, String> consumer1 = createKafkaConsumer(maxWaitMs, minBytes);
+        consumer1.assign(topicPartitions);
+        ConsumerRecords<String, String> emptyResult = consumer1.poll(Duration.ofMillis(maxWaitMs));
+        assertEquals(0, emptyResult.count());
+
+        ExecutorService executorService = Executors.newFixedThreadPool(1);
+
+        long startTime = System.currentTimeMillis();
+        AtomicLong endTime = new AtomicLong(0);
+        CountDownLatch latch = new CountDownLatch(1);
+        executorService.submit(() -> {
+            int totalRead = 0;
+            do {
+                ConsumerRecords<String, String> result = consumer1.poll(Duration.ofMillis(maxWaitMs));
+                result.forEach(record -> {
```

Comment:
This test is not accurate. There is a chance that `poll` is called after some messages are sent. Instead of testing a real world consumer, it would be better to test `KafkaRequestHandler#handleFetchRequest` so that we won't need to run a consumer in another thread.

Reply:
Yes, I'll add more tests in `KafkaRequestHandler#handleFetchRequest`.

For the new delayed fetch design, do you have any comments?

Reply:
No at the moment.

I just started reviewing this design. The first thing I'm concerned about is the test that shows the difference brought by this PR. Actually I wrote a similar unit test, before this PR it took about 1472 ms to receive all 10 messages. The time is far less than the max wait ms (3000 ms). So I found the test is somehow meaningless.

I'm still going to minimize the code changes. You don't have to address my comments before I complete the review. Maybe I'll open a new PR based on this PR, or just leave some comments or requested changes.

Reply:
The test:

```java
    @Test
    public void testFetchMinBytesSingleConsumer() throws Exception {
        final String topic = "testMinBytesTopic";
        admin.topics().createPartitionedTopic(topic, 1);

        final int maxWaitMs = 3000;
        final int minBytes = 1;

        @Cleanup
        final KafkaConsumer<String, String> consumer1 = createKafkaConsumer(maxWaitMs, minBytes);
        consumer1.assign(Collections.singleton(new TopicPartition(topic, 0)));
        assertTrue(consumer1.poll(Duration.ofMillis(maxWaitMs)).isEmpty());

        final ExecutorService executorService = Executors.newSingleThreadExecutor();
        final CountDownLatch latch = new CountDownLatch(1);
        final int numMessages = 10;
        final long startTime = System.currentTimeMillis();
        final AtomicLong endTime = new AtomicLong(0);
        executorService.execute(() -> {
            for (int i = 0; i < numMessages; ) {
                final int previous = i;
                for (ConsumerRecord<String, String> record : consumer1.poll(Duration.ofMillis(maxWaitMs))) {
                    log.info("Received record: {}", record.value());
                    i++;
                }
                log.info("The consumer read {} records totalRead {}", i - previous, i);
            }
            endTime.set(System.currentTimeMillis());
            latch.countDown();
        });

        final KafkaProducer<String, String> producer = createKafkaProducer();
        produceData(producer, Collections.singletonList(new TopicPartition(topic, 0)), numMessages);
        latch.await();
        log.info("Received {} records for {} ms", numMessages, endTime.get() - startTime);
        executorService.shutdown();
    }
```

The log:

> 15:52:47.022 [main:io.streamnative.pulsar.handlers.kop.KafkaApisTest@1067] INFO  io.streamnative.pulsar.handlers.kop.KafkaApisTest - Received 10 records for 1272 ms

Reply:
A more accurate test example:

```java
    @Test
    public void testFetchMinBytesSingleConsumer() throws Exception {
        final String topic = "testMinBytesTopic";
        final TopicPartition topicPartition = new TopicPartition(topic, 0);
        admin.topics().createPartitionedTopic(topic, 1);

        final int maxWaitMs = 3000;
        final int minBytes = 1;
        @Cleanup
        final KafkaHeaderAndRequest request = buildRequest(FetchRequest.Builder.forConsumer(maxWaitMs, minBytes,
                Collections.singletonMap(topicPartition, new FetchRequest.PartitionData(
                        0L, -1L, 1024 * 1024, Optional.empty()
                ))));
        final CompletableFuture<AbstractResponse> future = new CompletableFuture<>();
        final long startTime = System.currentTimeMillis();
        kafkaRequestHandler.handleFetchRequest(request, future);

        // Trigger the fetch
        final int numMessages = 10;
        final KafkaProducer<String, String> producer = createKafkaProducer();
        produceData(producer, Collections.singletonList(new TopicPartition(topic, 0)), numMessages);

        final FetchResponse<MemoryRecords> response = ((ResponseCallbackWrapper)
                future.get(maxWaitMs + 1000, TimeUnit.MILLISECONDS)).getResponse();
        final long endTime = System.currentTimeMillis();
        log.info("Take {} ms to process FETCH request, record count: {}",
                endTime - startTime, response.responseData().size());
    }
```

> 17:27:16.386 [main:io.streamnative.pulsar.handlers.kop.KafkaApisTest@1062] INFO  io.streamnative.pulsar.handlers.kop.KafkaApisTest - Take 3028 ms to process FETCH request, record count: 1